docs: Upgrade Kubernetes version (#1537)
redpanda-data/documentation: docs/manage/kubernetes/upgrade-kubernetes.mdx

---
title: Upgrade Kubernetes on Worker Nodes Running Redpanda
tags:
  - Kubernetes
deployment: Kubernetes
linkRoot: ../../../
description: Upgrading the Kubernetes version in a cluster ensures that your infrastructure is up to date and secure. This process involves updating the Kubernetes control plane and worker nodes to a new version while also making sure that the Redpanda cluster continues to function with minimal downtime.
---

<head>
    <meta name="title" content="Upgrade Kubernetes on Worker Nodes Running Redpanda | Redpanda Docs"/>

>     <meta name="description" content="Upgrading the Kubernetes version in a cluster ensures that your infrastructure is up to date and secure. This process involves updating the Kubernetes control plane and worker nodes to a new version while also making sure that the Redpanda cluster continues to function with minimal downtime."/>

Author: ```suggestion
    <meta name="description" content="Upgrading the Kubernetes version in a cluster ensures that your infrastructure is up to date and secure. This process involves updating the Kubernetes control plane and worker nodes to a new version while also making sure that the Redpanda cluster continues to function with minimal downtime."/>
    <link rel="canonical" href="https://docs.redpanda.com/docs/manage/kubernetes/upgrade-kubernetes/" />
```

## Prerequisites

- A staging environment in which to test the upgrade procedure before performing it in production.

- A running Redpanda deployment on a Kubernetes cluster.

- Familiarity with your hosting platform (GKE, EKS, AKS, or self-hosted) and any CLI tools your platform provides, such as `gcloud` or `eksctl`.

- Confirmation that the new version of Kubernetes is compatible with the version of Redpanda you are using. See the [Helm chart requirements](../../../reference/redpanda-helm-spec#requirements).

## Minimize data loss and downtime during Kubernetes upgrades

Upgrading the Kubernetes version involves updating Kubernetes components and other dependencies, which requires node restarts. To minimize data loss and downtime during the upgrade process, you must take precautions. This section outlines essential information on how the Helm chart helps maintain high availability and what steps you can take to avoid data loss, depending on the type of storage volumes that you use.

The Helm chart provides the following features to minimize data loss and downtime during the upgrade process:

> - A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process.

Author: ```suggestion
- A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process. For the default PDB, see [budget.maxUnavailable](../../../reference/redpanda-helm-spec) in the Helm chart.
```

> - A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process. For the default PDB, see [budget.maxUnavailable](../../../reference/redpanda-helm-spec) in the Helm chart.

Author: ```suggestion
- A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process. For the default PDB, see [`budget.maxUnavailable`](../../../reference/redpanda-helm-spec) in the Helm chart.
```

> - A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process. For the default PDB, see [`budget.maxUnavailable`](../../../reference/redpanda-helm-spec) in the Helm chart.

Author: ```suggestion
- A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process. For the default PDB, see [`budget.maxUnavailable`](../../../reference/redpanda-helm-spec#statefulsetbudgetmaxunavailable) in the Helm chart.
```

- Graceful shutdowns for all Pods within the StatefulSet to reduce potential data corruption.

Depending on your type of storage volume, you must take precautions to ensure your deployment can tolerate node restarts and avoid data loss:

### PersistentVolumes

If you use PersistentVolumes (PV) to store Redpanda data, take the following precautions:

- Verify compatibility between your storage classes and PersistentVolumes (PV) configurations with the new Kubernetes version.

> - If you use local-storage-backed PVs (such as Rancher Local Path Provisioner), your data is bound to the node where the PV is created, risking data loss during node deletions or failures. See [Best practices for local storage](#best-practices-for-local-storage).

Author: ```suggestion
- If you use local-storage-backed PVs (such as Rancher Local Path Provisioner), your data is bound to the node where the PV is created, risking data loss during node deletions or failures. To minimize the risk of data loss, see [Best practices for local storage](#best-practices-for-local-storage).
```

  Data in remote PVs remains safe during upgrades, as the PV is decoupled from the worker nodes.

### `hostPath`

> Data in `hostPath` volumes has a higher risk during upgrades because it is bound to a specific node. See [Best practices for local storage](#best-practices-for-local-storage).

Author: ```suggestion
Data in `hostPath` volumes is bound to a specific node, risking data loss during node deletions or failures. To minimize the risk of data loss, see [Best practices for local storage](#best-practices-for-local-storage).
```

> Data in `hostPath` volumes has a higher risk during upgrades because it is bound to a specific node. See [Best practices for local storage](#best-practices-for-local-storage).

Author: Added clarification.

### `emptyDir`

> Data in `emptyDir` volumes is ephemeral and will be lost when the container is terminated or the node is deleted. See [Best practices for local storage](#best-practices-for-local-storage).

Author: ```suggestion
Data in `emptyDir` volumes is ephemeral and will be lost when the container is terminated or the node is deleted. To minimize the risk of data loss, see [Best practices for local storage](#best-practices-for-local-storage).
```

The Helm chart provides the following features to minimize data loss and downtime during the upgrade process:

- A Pod Disruption Budget (PDB) that limits the number of concurrently unavailable Pods within the StatefulSet, ensuring the desired level of redundancy. Hosting platforms such as EKS and GKE respect the PDB during node upgrades, providing a safer upgrade process.

- Graceful shutdowns for all Pods within the StatefulSet to reduce potential data corruption.

Depending on your type of storage volume, you must take precautions to ensure your deployment can tolerate node restarts and avoid data loss:

- **PersistentVolume (PV)**: Verify compatibility between your storage classes and PV configurations with the new Kubernetes version.

  - **Remote PV**: Data in remote PVs remains safe during upgrades, as the PV is decoupled from the worker nodes.

  - **Local PV**: Data in local-storage-backed PVs (such as Rancher Local Path Provisioner) is tied to the node where the PV is created, risking data loss during node deletions or failures. See [Best practices for local storage](#best-practices-for-local-storage).

> - **`hostPath`**: Data in `hostPath` volumes has a higher risk during upgrades, as it is tied to a specific node. Migrate data to a PV or create backups before upgrading to avoid data loss. See [Best practices for local storage](#best-practices-for-local-storage).

Reviewer: Maybe `Best practices for local storage` could mention Tiered storage capability which could recover some segements.

Author: so if users set up remote writes, Redpanda will write old offsets to Tiered Storage and we can use that to recover if necessary?

- **`emptyDir`**: Data in `emptyDir` volumes is ephemeral and will be lost when the container is terminated or the node is deleted. Migrate data to a PV or create backups before upgrading to avoid data loss. See [Best practices for local storage](#best-practices-for-local-storage).

> ### Best practices for local storage

Reviewer: There is one problem with PersistentVolumeClaim. When user would drain Kubernetes node the Pod will be in Pending state forever as PersistentVolumeClaim is bounded to PersistentVolume which has node affinity. There need to be manual process of deleting PVC and deleting Pod. 

We could get away with deleting only PVC if draining would be run after that. The draining would be responsible for deleting our Pod, but due to a fact that PVC has already deletionTimestamp it should be recreated (I hope/didn't tested it).

P.S.
With Kubernetes 1.27 there will be small improvement https://github.com/kubernetes/kubernetes/pull/113270 where PersistentVolumeClaim would be recreated if Pod is in Pending state.

Author: I can add this?

> If you're running on Kubernetes version 1.26 or earlier, make sure to delete the PVC before upgrading the worker node. Otherwise, your Pods will remain in a pending state when they are rescheduled.

Reviewer: I agree, but please mention that PVC will stay until Pod is evicted by draining mechanism from Kubernetes. 

User can see this in metadata -> deletion timestamp.

When using local storage, data loss can occur during Kubernetes upgrades. To minimize the risk, follow these best practices before you upgrade:

> - **Back up data**: Create a backup of the data stored on the volumes. Backups ensure that you have a recovery option in case of data loss during the upgrade process.

Reviewer: I don't think we can create backups of running system. I would like to hear confirmation from core-storage team that if Redpanda is running operator could create a snapshot of the data folder which can then be used to start the same Redpanda instance from mentioned backup. What could happen with opened segments? Are they recoverable?

Author: maybe we'd need to put the broker in maintenance mode first? cc @jcsp

> - **Replicate all topics**: If you have topics with a replication factor of 1, temporarily increase the replication factor of those topics before you upgrade. A replication factor of at least 2 ensures that even if one node experiences data loss, you can still recover the data from a replica on another node. See [`rpk topic alter-config`](../../../reference/rpk/rpk-topic/rpk-topic-alter-config).

Reviewer: As far as I remember when the replication factor is changed one need to wait for under replicated metric to show 0 for all partitions for the altered topic.
